File: LeetCode/Problems/0002_Add_Two_Numbers/add_two_numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        return self.__recursiveHelper(l1, l2, 0)

    def __recursiveHelper(self, l1, l2, carry) -> ListNode:
        if l1 == None and l2 == None and carry == 0:
            return None

        x = l1.val if l1 != None else 0
        y = l2.val if l2 != None else 0
        sum = x + y + carry
        carry = sum // 10  # '//' is a floor division operator
        new_node = ListNode(sum % 10)

        l1 = l1.next if l1 != None else None
        l2 = l2.next if l2 != None else None
        new_node.next = self.__recursiveHelper(l1,l2,carry)

        return new_node

def printList(l) -> str:
    value = []
    while(l):
        value.append(l.val)
        l = l.next
    return('(' + ' -> '.join(map(str, value)) + ')')

l1 = ListNode(2, ListNode(4, ListNode(3)))
l2 = ListNode(5, ListNode(6, ListNode(4)))
print("Input: " + printList(l1) + " + " + printList(l2))

l3 = Solution().addTwoNumbers(l1, l2)
print("Output: " + printList(l3))

File: LeetCode/Problems/0002_Add_Two_Numbers/add_two_numbers.cpp
#include <iostream>

using namespace std;

// Definition for singly-linked list
struct ListNode{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution{
public:
    Solution(){}

    // time complexity: O(max(m,n)), space complexity: O(max(m,n))
    ListNode* addTwoNumber(ListNode* l1, ListNode* l2){
        // return recursive(l1, l2);
        return iterative(l1, l2);
    }

private:
    ListNode* recursive(ListNode* l1, ListNode* l2){
        return helper(l1, l2, 0);
    }

    ListNode* helper(ListNode* l1, ListNode* l2, int carry){
        if(!l1 && !l2 && !carry)
            return nullptr;

        ListNode dummyHead(0);
        ListNode *currNode = &dummyHead;  // to avoid potential memory leak

        int x = l1 ? l1->val : 0;
        int y = l2 ? l2->val : 0;
        int sum = x + y + carry;
        carry = sum / 10;

        ListNode *node = new ListNode(sum % 10);
        currNode->next = node;
        currNode = currNode->next;

        l1 = l1 ? l1->next : nullptr;
        l2 = l2 ? l2->next : nullptr;
        currNode->next = helper(l1, l2, carry);

        return dummyHead.next;
    }

    ListNode* iterative(ListNode* l1, ListNode* l2){
        int carry = 0;
        ListNode dummyHead(0);
        ListNode *currNode = &dummyHead;  // to avoid potential memory leak

        while(l1 || l2 || carry){
            int x = l1 ? l1->val : 0;
            int y = l2 ? l2->val : 0;
            int sum = x + y + carry;
            carry = sum / 10;

            ListNode *node = new ListNode(sum % 10);
            currNode->next = node;
            currNode = currNode->next;

            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }

        return dummyHead.next;
    }
};

string printList(ListNode* const head){
    string str;
    ListNode *curr = head;

    if(!curr)
        return str.append("Empty list");

    str.append("(");
    while(curr){
        str.append(std::to_string(curr->val));
        if(curr->next)
            str.append(" -> ");
        curr = curr->next;
    }
    str.append(")");

    return str;
}

void deleteList(ListNode** head){
    ListNode *curr = *head;
    ListNode *next;

    while(curr){
        next = curr->next;
        delete(curr);
        curr = next;
    }
    *head = nullptr;
}

int main(){

    ListNode *l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
    ListNode *l2 = new ListNode(5, new ListNode(6, new ListNode(4)));

    cout << "Input: " << printList(l1) << " + " << printList(l2) << "\n";

    Solution sol;
    ListNode *l3 = sol.addTwoNumber(l1, l2);

    cout << "Output: " << printList(l3) << "\n";

    deleteList(&l1);
    deleteList(&l2);
    deleteList(&l3);
    // cout << printList(l1) << "\n";
    // cout << printList(l2) << "\n";
    // cout << printList(l3) << "\n";

    return 0;
}
